pyveda/frameworks/augmentation_generator2.py

```diff
@@ -0,0 +1,97 @@
+import numpy as np
+
+from random import randint
+from random import sample
+
+from pyveda.frameworks.transforms import *
+
+
+class BaseGenerator():
+    def __init__(self, cache, shape=None, batch_size=32, shuffle=True, rescale_toa=False):
+        self.cache = cache
+        self.batch_size = batch_size
+        self.index = 0
+        self.shuffle = shuffle
+        self.rescale_toa = rescale_toa
+        self.on_epoch_end()
+
+    def build_batch(self, index):
+        raise NotImplemented
+
+    def __getitem__(self, index):
+        # yield self.build_batch(index)
+        return self.build_batch(index)
+
+    def __next__(self):
+        try:
+            item = self[self.index]  # index???
+        except IndexError:
+            raise StopIteration
+        self.index += 1
+        return item
+
+    def on_epoch_end(self):
+        '''update index for each epoch'''
+        # self.indexes = np.arange(len(self.list_ids)) self.list_ids not defined in baseclass
+        self.indexes = np.arange(len(self.cache))
+        if self.shuffle:
+            np.random.shuffle(self.indexes)
+
+    def __iter__(self):
+        """Create a generator that iterates over the Sequence."""
+        for item in (self[i] for i in range(len(self))):
+            self.index += 1
+            yield item
+
+    def __len__(self):
+        '''Denotes the number of batches per epoch'''
+        return int(np.floor(len(self.cache)/self.batch_size))
+
+
+class VedaStoreGenerator(BaseGenerator):
+    '''
+    VedaBase
+    '''
+    def __init__(self, cache, batch_size, rescale_toa=False):
+        super().__init__(cache, batch_size=batch_size, shuffle=True, rescale_toa=rescale_toa)
+        self.list_ids = np.arange(0, len(self.cache))
+        self.mltype = cache._trainer.mltype
+        self.shape = cache._trainer.image_shape
+
+    def build_batch(self, index):
+        '''Generate one batch of data'''
+        if index > len(self):
+            raise IndexError("index is invalid")
+        indexes = self.indexes[index*self.batch_size:(index+1)*self.batch_size]
+        list_ids_temp = [self.list_ids[k] for k in indexes]
+        X, y = self._data_generation(list_ids_temp)
+        return X, y
+
+    def _data_generation(self, list_ids_temp):
+        '''Generates data containing batch_size samples
+        optionally pre-processes the data'''
+
+        X = np.empty((self.batch_size, *self.shape[::-1]))  # issue with shape?
```

Comment:
Yes. For it to be in the `BaseClass` I think `VedaBase` and `VedaStream` will need to have `image_shape` and `mltype` in the same way, instead of `_vset.image_shape` vs. `_trainer.image_shape`, but there's a PR in for that change. Unless there's another way you think it should be implemented?

Reply:
moved to `BaseGenerator` class